src/Connector/CachingConnector.php

```diff
@@ -26,23 +28,48 @@ public function __construct(CacheItemPoolInterface $cache = null)
         $this->cache = $cache ?: new MemoryCache;
     }
 
-    public function fetch($source, EncapsulatedOptions $options = null)
-    {
+    /**
+     * @param string $source
+     * @param EncapsulatedOptions|null $options
+     * @param CacheKeyGeneratorInterface $cacheKeyGenerator
+     * @return mixed
+     * @throws InvalidCacheKeyException
+     */
+    public function fetch(
+        $source,
+        EncapsulatedOptions $options = null,
+        CacheKeyGeneratorInterface $cacheKeyGenerator = null
+    ) {
         $optionsCopy = $options ? $options->copy() : [];
 
+        $key = null;
+
         if ($this->isCacheEnabled()) {
             ksort($optionsCopy);
 
-            $hash = $this->hash([$source, $optionsCopy]);
+            if ($cacheKeyGenerator !== null) {
+                $key = $cacheKeyGenerator->generateCacheKey($source, $options);
+                if (!is_string($key)) {
+                    throw new InvalidCacheKeyException('Cache key must be of type string.');
+                }
+                if (strpbrk($key, '{}()/\@:') !== false) {
+                    throw new InvalidCacheKeyException(sprintf(
+                        'Cache key "%s" contains reserved characters {}()/\@:',
+                        $key
+                    ));
+                }
```

Comment:
This method is too big now. Remember this method is just called `fetch` and should be focused on data retrieval. Let's move this new code to a private method called `validateCacheKey`, since that seems to be what it's doing.